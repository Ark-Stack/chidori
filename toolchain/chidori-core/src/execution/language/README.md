# Language

## Goals
The intention here is to be able to construct reactive
graphs of computations to be executed by our runtime.
Specifically we want to provide reactive semantics to build a graph of computations.
This graph of computations is something that we can more cleanly visualize and reason about.
The reactivity of these computations is the boundary that we want to create between the
definition of a given program and the execution of that program.

## Taking control of execution

We want to be able to take control of the execution of a program in order to give
improved durability guarantees. This means that we want to be able to control the
execution of a program in order to be able to checkpoint the state of the program.

In instances where there are calls to our std api functions, we want to be able to delegate back our our
execution runtime. In order to do this and to instrument internal functions of the user's definitions - we mutate their
AST with a special "checkpoint" function call. This function call is then intercepted by our runtime and we can pause execution,
run our code, and then resume.


## Intermediate experiments

* Implementing our own language to define these graphs
* Providing an API for users to explicitly define these relationships
* Adopting another existing scripting language and manipulating its stack and heap
* Adopting WASM as our execution runtime and implementing our time travel features on top of it

From these experiments it feels clear that the best way to provide this functionality
is to make discovering these relationships implicit. 

These relationships also need to be identified at "compile" time, prior to the program running.
Otherwise the author needs to first think about how they'll structure their computation in two passes with
different entrypoints. This is a bad developer experience.

## The solution

To provide the best user experience and to provide the most flexibility in the future: we're simply implementing our own
initial pass over javascript and python ASTs. We manipulate the AST representation of the code base in order to identify relationships
across function blocks. We construct the graph of computations and then we execute the program ourselves.

This is signficiantly more simple than the other explored alternatives. We implement our own patterns for capturing values and constructing
snapshots of execution in the given languages.



## Unique features

This intersection allows us to also bridge multiple runtimes. We can declare a python function that depends on the execution
of a javascript function. We can declare an LLM prompt that depends on the execution of a python function.