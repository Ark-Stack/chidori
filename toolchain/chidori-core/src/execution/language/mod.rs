// This language exists to be able to author lazily evaluated functions.
// It's possible to do this in Rust, but it's not ergonomic.

pub mod compile;
pub mod javascript;
pub mod parser;
pub mod python;
