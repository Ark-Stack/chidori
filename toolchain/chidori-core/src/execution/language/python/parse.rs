use rustpython_parser::{ast, Parse};

fn example() {
    let python_source = r#"
def is_odd(i):
  return bool(i & 1)
"#;
    let ast = ast::Suite::parse(python_source, "<embedded>");

    assert!(ast.is_ok());
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;

    #[test]
    fn test_evaluation_single_node() {
        let python_source = indoc! { r#"
            def bar():
               print("bar")
            
            def foo():
                test = prompt.split("\n")
                print("Hello World!")
            "#};
        let ast = ast::Suite::parse(python_source, "<embedded>").unwrap();

        for item in ast {
            match item {
                ast::Stmt::FunctionDef(ast::StmtFunctionDef { name, .. }) => {
                    dbg!(name);
                }
                _ => {}
            }
        }
    }
}
