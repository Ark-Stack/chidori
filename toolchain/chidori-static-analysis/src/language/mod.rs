// This language exists to be able to author lazily evaluated functions.
// It's possible to do this in Rust, but it's not ergonomic.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::hash::{Hash, Hasher};
use petgraph::graph::DiGraph;
use petgraph::graphmap::DiGraphMap;
use petgraph::visit::EdgeRef;
use thiserror::Error;

pub mod typechecker;
pub mod javascript;
pub mod python;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Eq, Hash)]
pub struct TextRange {
    pub start: usize,
    pub end: usize,
}

// TODO: implement a function that infers the language from the source code successfully parsing

// TODO: it would be helpful if reports noted if a value is a global, an arg, or a kwarg
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ReportItem {
    // pub context_path: Vec<ContextPath>,
}

#[derive(Debug, Default, Clone, PartialEq, Serialize, Deserialize)]
pub struct ReportTriggerableFunctions {
    pub arguments: Vec<String>,
    // pub context_path: Vec<ContextPath>,
    // TODO: these need their own set of depended values
    // TODO: we need to extract signatures for triggerable functions
    pub emit_event: Vec<String>,
    pub trigger_on: Vec<String>,
}

#[derive(Debug, Default, Clone)]
pub struct InternalCallGraph {
    graph: DiGraph<String, ()>
}

impl Serialize for InternalCallGraph {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where
            S: serde::Serializer,
    {
        let mut adjacency_list = HashMap::new();
        for node in self.graph.node_indices() {
            let edges = self.graph.edges(node)
                .map(|edge| edge.target().index())
                .collect::<Vec<_>>();
            adjacency_list.insert(self.graph[node].clone(), edges);
        }
        adjacency_list.serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for InternalCallGraph {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
        where
            D: serde::Deserializer<'de>,
    {
        let adjacency_list: HashMap<String, Vec<usize>> = HashMap::deserialize(deserializer)?;
        let mut graph = DiGraph::new();
        let mut node_indices = HashMap::new();

        // Add nodes
        for node in adjacency_list.keys() {
            let index = graph.add_node(node.clone());
            node_indices.insert(node.clone(), index);
        }

        // Add edges
        for (node, edges) in adjacency_list {
            let source_index = node_indices[&node];
            for target_index in edges {
                if let Some(&target_node_index) = node_indices.values().find(|&&idx| idx.index() == target_index) {
                    graph.add_edge(source_index, target_node_index, ());
                }
            }
        }

        Ok(InternalCallGraph { graph })
    }
}

impl PartialEq for InternalCallGraph {
    fn eq(&self, other: &Self) -> bool {
        // Compare the graphs for equality
        self.graph.node_indices().all(|n| {
            self.graph[n] == other.graph[n]
        }) && self.graph.edge_indices().all(|e| {
            self.graph.edge_endpoints(e) == other.graph.edge_endpoints(e)
        })
    }
}

impl Eq for InternalCallGraph {}

impl Hash for InternalCallGraph {
    fn hash<H: Hasher>(&self, state: &mut H) {
        // Hash the nodes and edges of the graph
        for node in self.graph.node_indices() {
            self.graph[node].hash(state);
        }
        for edge in self.graph.edge_indices() {
            self.graph.edge_endpoints(edge).hash(state);
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Report {
    pub internal_call_graph: InternalCallGraph,
    pub cell_exposed_values: HashMap<String, ReportItem>,
    pub cell_depended_values: HashMap<String, ReportItem>,
    pub triggerable_functions: HashMap<String, ReportTriggerableFunctions>,
}


#[derive(Debug, Clone, Eq, Hash, PartialEq, Serialize, Deserialize)]
pub enum ContextPath {
    Initialized,
    InFunction(String, TextRange),
    InAnonFunction,
    FunctionArguments,
    FunctionArgument(String),
    InClass(String),
    InFunctionDecorator(usize),
    InCallExpression,
    ChName,
    AssignmentToStatement,
    AssignmentFromStatement,
    // bool = true (is locally defined)
    IdentifierReferredTo {
        name: String,
        in_scope: bool,
        exposed: bool
    },
    Attribute(String),
    Constant(String),
}


#[derive(Error, Debug, Serialize, Deserialize)]
pub enum ChidoriStaticAnalysisError {
    #[error("Unknown chidori analysis error")]
    Unknown,
    #[error("Parse error at offset {offset} in {source_path}: {msg}. Source: {source_code}")]
    ParseError {
        msg: String,
        offset: u32,
        source_path: String,
        source_code: String
    },
}



