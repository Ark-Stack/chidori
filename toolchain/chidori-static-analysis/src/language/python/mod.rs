pub mod parse;
mod flow;
