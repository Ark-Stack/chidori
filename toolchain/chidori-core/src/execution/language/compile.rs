use crate::execution::execution::execution_graph::ExecutionGraph;
use crate::execution::execution::execution_state::ExecutionState;
/// Logic to convert the rust ast to our scheduler graph
use crate::execution::language::parser::{BinaryOp, Error, Expr, Func, Program, Value};
use std::collections::HashMap;

use crate::execution::primitives::serialized_value::{
    serialize_to_vec, RkyvSerializedValue as RSV,
};

// TODO: how does this actually fetch the functions?
fn compile_to_graph(program: Program) -> ExecutionGraph {
    let mut db = ExecutionGraph::new();
    let mut state = ExecutionState::new();
    let state_id = (0, 0);

    // We start with the number 0 at node 0
    let (state_id, mut state) = db.upsert_operation(
        state_id,
        state,
        0,
        0,
        Box::new(|_args| {
            let v = RSV::Number(0);
            return serialize_to_vec(&v);
        }),
    );

    if let Some(main_func) = program.funcs.get("main") {
        eval_to_graph(&main_func.body.0, &program.funcs, &mut db);
    }
    // ast.body

    // fn eval_expr(
    //     expr: &Spanned<Expr>,
    //     funcs: &HashMap<String, Func>,
    //     stack: &mut Vec<(String, Value)>,
    // ) -> Result<Value, Error> {

    // TODO: look for a "main" function
    // 1. For each function call, create a node
    // 2. For each import, create a node
    // 3. Create connections between nodes when we refer to variables
    // 4. For each function assigned to a variable, create a connection to all function invocations that refer to that variable
    db
}

/// This walks the AST and constructs a graph of operations
fn eval_to_graph(
    expr: &Expr,
    funcs: &HashMap<String, Func>,
    db: &mut ExecutionGraph,
) -> Result<(), Error> {
    // match &expr {
    //     Expr::Error => unreachable!(), // Error expressions only get created by parser errors, so cannot exist in a valid AST
    //     Expr::Value(val) => val.clone(),
    //     Expr::List(items) => Value::List(),
    //     Expr::Local(name) => stack,
    //     Expr::Let(local, val, body) => {}
    //     Expr::Then(a, b) => {
    //         eval_to_graph(a, funcs, stack)?;
    //         eval_to_graph(b, funcs, stack)?
    //     }
    //     Expr::Binary(a, BinaryOp::Add, b) => Value::Num(),
    //     Expr::Binary(a, BinaryOp::Sub, b) => Value::Num(),
    //     Expr::Binary(a, BinaryOp::Mul, b) => Value::Num(),
    //     Expr::Binary(a, BinaryOp::Div, b) => Value::Num(),
    //     Expr::Binary(a, BinaryOp::Eq, b) => {}
    //     Expr::Binary(a, BinaryOp::NotEq, b) => {}
    //     Expr::Binary(a, BinaryOp::PipeOp, b) => {}
    //     Expr::Call(func, args) => {}
    //     Expr::If(cond, a, b) => {}
    //     Expr::Print(a) => {}
    // }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::execution::language::parser::parse;
    use std::collections::HashMap;

    #[test]
    fn test_compiling_simple_program() {
        let program = parse(
            r#"
        fn main() {
            let x = 1;
            let y = 2;
            print(x + y);
            x |>
            y |>
            print(x + y);
        }
    "#
            .to_string(),
        )
        .unwrap()
        .unwrap();

        let db = compile_to_graph(program);
    }
}
