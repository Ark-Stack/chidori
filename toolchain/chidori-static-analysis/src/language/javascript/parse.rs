extern crate swc_ecma_parser;

use crate::language::python;
use rustpython_parser::ast::{Constant, Identifier};
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use swc_common::sync::Lrc;
use swc_common::{
    errors::{ColorConfig, Handler},
    FileName, FilePathMapping, SourceMap,
};
use swc_ecma_ast as ast;
use swc_ecma_ast::{Decl, Expr, FnDecl, ModuleDecl, ModuleItem, Pat, PatOrExpr, Stmt};
use swc_ecma_parser::{lexer::Lexer, Parser, StringInput, Syntax};

#[derive(Debug, Clone, Eq, Hash, PartialEq, Serialize, Deserialize)]
pub enum ContextPath {
    Initialized,
    InFunction(String),
    InFunctionDecorator(usize),
    InCallExpression,
    ChName,
    AssignmentToStatement,
    AssignmentFromStatement,
    // bool = true (is locally defined)
    IdentifierReferredTo(String, bool),
    Attribute(String),
    Constant(String),
}

#[derive(Debug, Clone, Default, PartialEq)]
pub struct FunctionContext {
    dependencies: Vec<Vec<ContextPath>>,
}

// TODO: we're building a filtered subset of the AST
// TODO: what we can consider is evaluating the ast into the references
// TODO: the context path is the accumulated state, when we pop it we evaluate it
#[derive(Default)]
pub struct ASTWalkContext {
    pub context_stack_references: Vec<Vec<ContextPath>>,
    pub context_stack: Vec<ContextPath>,
    pub locals: HashSet<String>,
    pub local_contexts: Vec<HashSet<String>>,
    pub globals: HashSet<String>,
}

impl ASTWalkContext {
    fn new() -> Self {
        Self {
            context_stack_references: vec![],
            context_stack: vec![],
            locals: HashSet::new(),
            local_contexts: vec![],
            globals: HashSet::new(),
        }
    }

    fn var_exists(&self, name: &str) -> bool {
        if self.globals.contains(name) {
            return true;
        }
        if self.locals.contains(name) {
            return true;
        }
        for local_context in self.local_contexts.iter().rev() {
            if local_context.contains(name) {
                return true;
            }
        }
        return false;
    }

    fn new_local_context(&mut self) {
        self.local_contexts.push(self.locals.clone());
    }

    fn pop_local_context(&mut self) {
        self.local_contexts.pop();

        // Clearing the existing locals
        self.locals.clear();

        // Union of all remaining hashsetss in local_contexts
        self.locals.extend(
            self.local_contexts
                .iter()
                .flat_map(|context| context.iter().cloned()),
        );
    }

    fn enter_statement_function(&mut self, name: &ast::Ident) -> usize {
        self.context_stack
            .push(ContextPath::InFunction(name.to_string()));
        self.context_stack_references
            .push(self.context_stack.clone());
        self.context_stack.len()
    }

    fn enter_decorator_expression(&mut self, idx: &usize) -> usize {
        self.context_stack
            .push(ContextPath::InFunctionDecorator(idx.clone()));
        self.context_stack.len()
    }

    fn enter_call_expression(&mut self) -> usize {
        self.context_stack.push(ContextPath::InCallExpression);
        self.context_stack.len()
    }

    fn encounter_constant(&mut self, name: &Constant) {
        if let Constant::Str(s) = name {
            self.context_stack
                .push(ContextPath::Constant(s.to_string()));
            self.context_stack_references
                .push(self.context_stack.clone());
        }
    }

    fn encounter_named_reference(&mut self, name: &Identifier) {
        // TODO: we need to check if this is a local variable or not
        if self.var_exists(&name.to_string()) {
            // true, the var exists in the local or global scope
            self.context_stack
                .push(ContextPath::IdentifierReferredTo(name.to_string(), true));
        } else {
            if let Some(ContextPath::AssignmentToStatement) = self.context_stack.last() {
                self.locals.insert(name.to_string());
            }
            self.context_stack
                .push(ContextPath::IdentifierReferredTo(name.to_string(), false));
        }
        self.context_stack_references
            .push(self.context_stack.clone());
        self.context_stack.pop();
    }

    fn enter_assignment_to_statement(&mut self) -> usize {
        self.context_stack.push(ContextPath::AssignmentToStatement);
        self.context_stack.len()
    }

    fn enter_assignment_from_statement(&mut self) -> usize {
        self.context_stack
            .push(ContextPath::AssignmentFromStatement);
        self.context_stack.len()
    }

    fn enter_attr(&mut self, name: &Identifier) -> usize {
        self.context_stack
            .push(ContextPath::Attribute(name.to_string()));
        self.context_stack.len()
    }

    fn pop_until(&mut self, size: usize) {
        while self.context_stack.len() >= size {
            self.context_stack.pop();
        }
    }
}

fn traverse_module(module: ModuleItem, machine: &mut ASTWalkContext) {
    match module {
        ModuleItem::ModuleDecl(mod_decl) => match mod_decl {
            ModuleDecl::Import(ast::ImportDecl { .. }) => {}
            ModuleDecl::ExportDecl(_) => {}
            ModuleDecl::ExportNamed(_) => {}
            ModuleDecl::ExportDefaultDecl(_) => {}
            ModuleDecl::ExportDefaultExpr(_) => {}
            ModuleDecl::ExportAll(_) => {}
            ModuleDecl::TsImportEquals(_) => {}
            ModuleDecl::TsExportAssignment(_) => {}
            ModuleDecl::TsNamespaceExport(_) => {}
        },
        ModuleItem::Stmt(stmt) => {
            let v = vec![stmt];
            traverse_stmts(v.as_slice(), machine);
        }
    }
}

fn traverse_expr(expr: &ast::Expr, machine: &mut ASTWalkContext) {
    match expr {
        Expr::This(ast::ThisExpr { .. }) => {}
        Expr::Array(ast::ArrayLit { .. }) => {}
        Expr::Object(ast::ObjectLit { .. }) => {}
        Expr::Fn(ast::FnExpr { .. }) => {}
        Expr::Unary(ast::UnaryExpr { .. }) => {}
        Expr::Update(ast::UpdateExpr { .. }) => {}
        Expr::Bin(ast::BinExpr { .. }) => {}
        Expr::Assign(ast::AssignExpr {
            op, left, right, ..
        }) => {
            let idx = machine.enter_assignment_to_statement();
            match left {
                PatOrExpr::Expr(expr) => {
                    traverse_expr(expr, machine);
                }
                PatOrExpr::Pat(pat) => {}
            }
            machine.pop_until(idx);
            traverse_expr(right, machine);
            machine.pop_until(idx);
        }
        Expr::Member(ast::MemberExpr { .. }) => {}
        Expr::SuperProp(ast::SuperPropExpr { .. }) => {}
        Expr::Cond(ast::CondExpr { .. }) => {}
        Expr::Call(ast::CallExpr { .. }) => {}
        Expr::New(ast::NewExpr { .. }) => {}
        Expr::Seq(ast::SeqExpr { .. }) => {}
        Expr::Ident(ast::Ident { .. }) => {}
        Expr::Lit(_) => {}
        Expr::Tpl(ast::Tpl { .. }) => {}
        Expr::TaggedTpl(ast::TaggedTpl { .. }) => {}
        Expr::Arrow(ast::ArrowExpr { .. }) => {}
        Expr::Class(ast::ClassExpr { .. }) => {}
        Expr::Yield(ast::YieldExpr { .. }) => {}
        Expr::MetaProp(ast::MetaPropExpr { .. }) => {}
        Expr::Await(ast::AwaitExpr { .. }) => {}
        Expr::Paren(ast::ParenExpr { .. }) => {}
        Expr::JSXMember(ast::JSXMemberExpr { .. }) => {}
        Expr::JSXNamespacedName(ast::JSXNamespacedName { .. }) => {}
        Expr::JSXEmpty(ast::JSXEmptyExpr { .. }) => {}
        Expr::JSXElement(el) => {}
        Expr::JSXFragment(ast::JSXFragment { .. }) => {}
        Expr::TsTypeAssertion(ast::TsTypeAssertion { .. }) => {}
        Expr::TsConstAssertion(ast::TsConstAssertion { .. }) => {}
        Expr::TsNonNull(ast::TsNonNullExpr { .. }) => {}
        Expr::TsAs(ast::TsAsExpr { .. }) => {}
        Expr::TsInstantiation(ast::TsInstantiation { .. }) => {}
        Expr::TsSatisfies(ast::TsSatisfiesExpr { .. }) => {}
        Expr::PrivateName(ast::PrivateName { .. }) => {}
        Expr::OptChain(ast::OptChainExpr { .. }) => {}
        Expr::Invalid(ast::Invalid { .. }) => {}
    }
}

fn traverse_stmts(stmts: &[Stmt], machine: &mut ASTWalkContext) {
    for stmt in stmts {
        match stmt {
            Stmt::Expr(expr_stmt) => {
                traverse_expr(&*expr_stmt.expr, machine);
            }
            Stmt::Block(block_stmt) => {
                traverse_stmts(&block_stmt.stmts, machine);
            }
            Stmt::Empty(_) => {}
            Stmt::Debugger(ast::DebuggerStmt { .. }) => {}
            Stmt::With(ast::WithStmt { .. }) => {}
            Stmt::Return(ast::ReturnStmt { .. }) => {}
            Stmt::Labeled(ast::LabeledStmt { .. }) => {}
            Stmt::Break(ast::BreakStmt { .. }) => {}
            Stmt::Continue(ast::ContinueStmt { .. }) => {}
            Stmt::If(ast::IfStmt { .. }) => {}
            Stmt::Switch(ast::SwitchStmt { .. }) => {}
            Stmt::Throw(ast::ThrowStmt { .. }) => {}
            Stmt::Try(t) => {
                let ast::TryStmt { block, .. } = &**t;
            }
            Stmt::While(ast::WhileStmt { .. }) => {}
            Stmt::DoWhile(ast::DoWhileStmt { .. }) => {}
            Stmt::For(ast::ForStmt { .. }) => {}
            Stmt::ForIn(ast::ForInStmt { .. }) => {}
            Stmt::ForOf(ast::ForOfStmt { .. }) => {}
            Stmt::Decl(decl) => match decl {
                Decl::Class(_) => {}
                Decl::Fn(ast::FnDecl {
                    ident, function, ..
                }) => {
                    let idx = machine.enter_statement_function(ident);
                    let ast::Function { body, .. } = &**function;
                    if let Some(body) = body {
                        traverse_stmts(&body.stmts, machine);
                    }
                    machine.pop_until(idx);
                }
                Decl::Var(_) => {}
                Decl::Using(_) => {}
                Decl::TsInterface(_) => {}
                Decl::TsTypeAlias(_) => {}
                Decl::TsEnum(_) => {}
                Decl::TsModule(_) => {}
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_evaluation_single_node() {
        let machine = &mut ASTWalkContext::new();
        let cm: Lrc<SourceMap> = Default::default();
        let handler = Handler::with_tty_emitter(ColorConfig::Auto, true, false, Some(cm.clone()));

        // Real usage
        // let fm = cm
        //     .load_file(Path::new("test.js"))
        //     .expect("failed to load test.js");
        let fm = cm.new_source_file(
            FileName::Custom("test.js".into()),
            r#"
            function bar() {
               console.log("bar");
            }
            
            function x() {
              ch.get("testing")
            }
            
            function x() {
              ch.set("testing", "other")
            }
            
            function v() {
            
            }
            
            function foo() {
                ch.useEffect(() => {
                    let test = prompt.split("\n");
                    console.log("Hello World!");
                }, ["bar"])
            }
            
            "#
            .into(),
        );
        let lexer = Lexer::new(
            // We want to parse ecmascript
            Syntax::Es(Default::default()),
            // EsVersion defaults to es5
            Default::default(),
            StringInput::from(&*fm),
            None,
        );

        let mut parser = Parser::new_from(lexer);

        for e in parser.take_errors() {
            e.into_diagnostic(&handler).emit();
        }

        let module = parser
            .parse_module()
            .map_err(|mut e| {
                // Unrecoverable fatal error occurred
                e.into_diagnostic(&handler).emit()
            })
            .expect("failed to parser module");

        for item in module.body {
            if let ModuleItem::Stmt(stmt) = item {
                match stmt {
                    Stmt::Decl(Decl::Fn(FnDecl {
                        ident, function, ..
                    })) => {
                        println!("Function name: {}", ident.sym);
                        println!("Function body: {:?}", function.body);
                        if let Some(body) = &function.body {
                            traverse_stmts(&body.stmts, machine);
                        }
                    }
                    _ => {}
                }
            }
        }
    }
}
