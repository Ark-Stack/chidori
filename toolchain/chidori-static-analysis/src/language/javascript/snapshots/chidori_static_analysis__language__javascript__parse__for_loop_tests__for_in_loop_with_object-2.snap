---
source: chidori-static-analysis/src/language/javascript/parse.rs
description: "const obj = { a: 1, b: 2, c: 3 };\nfor (const key in obj) {\n    console.log(key, obj[key]);\n}\n"
---
internal_call_graph: {}
cell_exposed_values:
  obj: {}
cell_depended_values: {}
triggerable_functions: {}
